### PR TITLE
add basic handler implementations

### PR DESCRIPTION
**Pull request by LemmaEOF**
- remove EnergyType from EnergyComponent
- add EnergyHandler and EnergySerializer for basic reference impls of handling

### CODE REVIEW

File: src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		return maxEnergy;
+	}
+
+	@Override
+	public int getCurrentEnergy() {
+		return currentEnergy;
+	}
+
+	@Override
+	public boolean canInsertEnergy() {
```
Comment (falkreon) [inline src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java:32]: I wonder if we shouldn't just consider canInsert / canExtract to be true unless specifically denied? We can always find out if there's room by asking getCurrent and getMax, or by simulating an insert.

File: src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+	@Override
+	public boolean canExtractEnergy() {
+		return currentEnergy > 0;
+	}
+
+	@Nonnull
+	@Override
+	public int insertEnergy(int amount, ActionType actionType) {
+		int insertRoom = maxEnergy = currentEnergy;
```
Comment (falkreon) [inline src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java]: This line probably wasn't what was intended

**Review by falkreon — COMMENTED** [review]
Now might be a good time to add JUnit and learn to build test cases, if you're up for it.

**Comment by LemmaEOF** [comment]
I don't know really anything about JUnit or test cases so I don't think I can at the moment.

**Comment by falkreon** [comment]
Alright, now we just need the handler to implement Observable and we're all set.

File: src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java [hunk 3]
```diff
@@ -60,4 +61,9 @@ void setMaxEnergy(int amount) {
 		maxEnergy = amount;
 		if (currentEnergy > maxEnergy) currentEnergy = maxEnergy;
 	}
+
+	@Override
+	public void listen(@Nonnull Runnable r) {
```
Comment (falkreon) [inline src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java:79]: Okay but we always need to call this if we change something.

**Review by falkreon — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java (hunk 3 above)
Comment (LemmaEOF) [reply src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java:79]: What do we call it with? We'd need to supply a Runnable.

Comment (falkreon) [reply src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java:79]: Look at the Concrete classes. They all have an onChanged(), which calls the cached Runnable. Any method in which the Component changes, calls onChanged on itself

File: src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java [hunk 4]
```diff
@@ -49,21 +53,28 @@ public int insertEnergy(int amount, ActionType actionType) {
 	public int extractEnergy(int amount, ActionType actionType) {
 		int extractAmount = (amount <= currentEnergy)? amount : currentEnergy;
 		if (actionType == ActionType.PERFORM) currentEnergy -= extractAmount;
+		onChanged();
```
Comment (falkreon) [inline src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java]: Let's only call onChanged though if it actually changed

File: src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java [hunk 5]
```diff
@@ -41,6 +44,7 @@ public int insertEnergy(int amount, ActionType actionType) {
 		int insertRoom = maxEnergy - currentEnergy;
 		int insertAmount = (amount <= insertRoom)? amount : insertRoom;
 		if (actionType == ActionType.PERFORM) currentEnergy += insertAmount;
+		onChanged();
```
Comment (falkreon) [inline src/main/java/io/github/cottonmc/energy/impl/EnergyHandler.java]: Only if changed

**Review by falkreon — CHANGES_REQUESTED** [review]

**Review by falkreon — APPROVED** [review]

**Pull request merged**